OpenRA.Mods.Common/Traits/RevealsMap.cs

```diff
@@ -0,0 +1,78 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Reveals shroud and fog across the whole map while active.")]
+	public class RevealsMapInfo : ConditionalTraitInfo
+	{
+		[Desc("Stance the watching player needs to see the shroud removed.")]
+		public readonly Stance ValidStances = Stance.Ally;
+
+		[Desc("Can this actor reveal shroud generated by the `GeneratesShroud` trait?")]
+		public readonly bool RevealGeneratedShroud = true;
+
+		public override object Create(ActorInitializer init) { return new RevealsMap(init.Self, this); }
+	}
+
+	public class RevealsMap : ConditionalTrait<RevealsMapInfo>, INotifyKilled, INotifyActorDisposing
+	{
+		readonly RevealsMapInfo info;
+		readonly Shroud.SourceType type;
+
+		public RevealsMap(Actor self, RevealsMapInfo info)
+			: base(info)
+		{
+			this.info = info;
+			type = info.RevealGeneratedShroud ? Shroud.SourceType.Visibility
+				: Shroud.SourceType.PassiveVisibility;
+		}
+
+		protected void AddCellsToPlayerShroud(Actor self, Player p, PPos[] uv)
+		{
+			if (!info.ValidStances.HasStance(p.Stances[self.Owner]))
+				return;
+
+			p.Shroud.AddSource(this, type, uv);
```

Comment:
Why isn't this using `ExploreAll`? Using `AddSource` looks like generating a lot of overhead.

Reply:
`Shroud.ExploreAll()` reveals only Shroud not Fog.

Reply:
I see, my bad. It might make sense to expose a method to do this.